### DISCUSSION TITLE
Ways to improve user experience

### DISCUSSION THREAD

**Discussion by DanielVernall** [dsa-ou/allowed#27] [Ideas]
Hi all,

In M269 I found allowed.py to be useful, but a bit finnicky to use on Windows. For example, `IPython` is not added to the Windows PATH environment variable when installing Anaconda, so when running `python allowed.py` in a terminal/cmd window it was not found, causing cells with magic included to be skipped (not great when TMA 03 had `%run -i` in lots of cells). So, I ended up reinstalling `IPython` via `pip` to ensure it was accessible in a terminal.

I also found that it was a bit awkward adding the path to the TMA files to the command, as I kept my TMA notebooks in separate folders and had to navigate to them outside of the terminal and copy and paste their location. Doable, but a bit long-winded.

So, I was thinking of ways to improve the user experience for the tool. I had two ideas:

1. Write a [Jupyter Notebook Extension](https://jupyter-notebook.readthedocs.io/en/stable/extending/contents.html)
    - We could run this [on file save](https://jupyter-notebook.readthedocs.io/en/stable/extending/savehooks.html) to check code whilst editing the notebook - avoiding students spending time on a "doomed" approach.
    - We could style the code cells in response to the output of `allowed.py`, potentially highlighting/underlining disallowed constructs.

2. Alternatively, add a simple HTML user-interface, with a file input control to easily browse and select the TMA folder/file.
    - We could create a very simple HTTP API using [Flask](https://flask.palletsprojects.com/en/2.0.x/quickstart/) which calls `allowed.py` functionality.
    - If we use this approach it might be easiest to bundle the python code (python, IPython and pytype) into a single binary using [PyInstaller](https://pyinstaller.org/en/stable/), or similar.

Let me know your thoughts,
Dan

**Comment by mwermelinger** [comment]
Dan, thanks for reporting those issues and for your suggestions. 

For next year, we're planning to use Python virtual environments and use a script that automates the installation and use of `allowed` so that the problems you mention should not occur (famous last words :-). Have a look at https://dsa-ou.github.io/m269-installer. The script creates an alias so that you can go to any folder where your TMA is and just type `allowed
TMA01.ipynb` (or whatever the file name). The alias gets expanded into `python path/to/allowed.py -c path/to/m269.json TMA01.ipynb`. 

A notebook extension that highlights the disallowed constructs sounds nice, so go ahead with that if you have the time and inclination. I don't think the rest of the team has the knowledge or time to invest on that. The problem is that the extension might work for one Jupyter UI, say JupyterLab, and not another, say VS Code or Jupyter Classic, so it's only worth investing time if it works for several UIs. Moreover, Jupyter Classic is being deprecated. 

Bundling sounds like a good idea too, but it's probably better to invest on making `allowed` pip' able, which would also install its dependencies.

**Comment by densnow** [comment]
@DanielVernall Thanks for sharing your ideas and feedback. 

The jupyter plugin does sound good, so if you do decide to have a crack at it (taking into account @mwermelinger 's comments about multiple UI's time and experience etc) then I would be willing to chip in. I don't have experience of Jupyter extensions myself, but I can always read up some things. Also there isn't really any need to rush here, so if you think it could be a good project to learn and practice some skills on then it might be worth doing.

When it comes to some of the issues you have been experiencing on windows I think (like has been said above) that enabling a `pip` installation of `allowed` could sort out many of them. It feels somewhat unfinished if we are invoking a command line tool with `python path/to/python_script.py`. We need to have it installed, and `pip` seems to be a popular and easy way to get this done.

With regards to adding the path to your TMA notebook to the command in the terminal, does windows not have tab-completion in the terminal (i.e. start typing, then press TAB to autocomplete)? If not, that does sound like a nightmare, and all that exact typing would certainly put people off!
